pkg/utils/applicator/applicator.go

```diff
@@ -0,0 +1,220 @@
+	UpdatingApply(context.Context, runtime.Object, ...ApplyOption) error
+}
+
+// UpdatingApplyFn implements UpdatingApplicator.
+type UpdatingApplyFn func(context.Context, runtime.Object, ...ApplyOption) error
+
+// UpdatingApply applies changes to the supplied object.
+func (fn UpdatingApplyFn) UpdatingApply(ctx context.Context, o runtime.Object, ao ...ApplyOption) error {
+	return fn(ctx, o, ao...)
+}
+
+// ApplyOption is called before applying the existing object to match the
+// desired object. ApplyOptions are still called even if the object does NOT exist.
+// If the object does not exist, `existing` will be assigned as `nil`.
+type ApplyOption func(ctx context.Context, existing, desired runtime.Object) error
+
+// NewAPIApplicator creates returns an Applicator that applies changes to an
+// object by either creating or updating/patching it in a Kubernetes API server.
+func NewAPIApplicator(c client.Client) *APIApplicator {
+	return &APIApplicator{creatorFn(createOrGetExisting), c}
+}
+
+// APIApplicator implements Applicator
+type APIApplicator struct {
+	creator
+	c client.Client
+}
+
+// UpdatingApply applies changes to an object with updating
+func (a *APIApplicator) UpdatingApply(ctx context.Context, desired runtime.Object, ao ...ApplyOption) error {
+	existing, err := a.createOrGetExisting(ctx, a.c, desired, ao...)
+	if err != nil {
+		return err
+	}
+	if existing == nil {
+		// the object does not exist
+		// create successfully
+		return nil
+	}
+
+	// the object already exists, update it
+	if err := applyApplyOptions(ctx, existing, desired, ao); err != nil {
+		return err
+	}
+	desired.(Object).SetResourceVersion(existing.(Object).GetResourceVersion())
+	return errors.Wrap(a.c.Update(ctx, desired), "cannot update object")
+}
+
+// PatchingApply applies changes to an object with patching
+func (a *APIApplicator) PatchingApply(ctx context.Context, desired runtime.Object, ao ...ApplyOption) error {
+	existing, err := a.createOrGetExisting(ctx, a.c, desired, ao...)
+	if err != nil {
+		return err
+	}
+	if existing == nil {
+		// the object does not exist before
+		// now it's created successfully
+		return nil
+	}
+
+	// the object already exists, patch it
+	if err := applyApplyOptions(ctx, existing, desired, ao); err != nil {
+		return err
+	}
+	var patch client.Patch
+	owner := meta.GetControllerOf(existing.(Object))
+	if owner == nil {
+		// fall back to use client-side apply
+		goto ClientSidePatch
+	}
+	// use server-side apply as first priority
+	patch = client.Apply
+	err = a.c.Patch(ctx, desired, patch, []client.PatchOption{client.FieldOwner(owner.UID), client.ForceOwnership}...)
+	if err != nil && isIncompatibleServerError(err) {
+		goto ClientSidePatch
+	}
+	if err != nil {
+		return errors.Wrap(err, "cannot patch object")
+	}
+	return nil
+ClientSidePatch:
+	// fall back to client-side patch if server-side not work
+	patch = client.Merge
+	return errors.Wrap(a.c.Patch(ctx, desired, patch), "cannot patch object")
```

Comment:
Usually We don't like `goto` in such case, you can write it in one line instead of using `goto` if you feel redundant.

```
return errors.Wrap(a.c.Patch(ctx, desired, client.Merge), "cannot patch object")
```